internal/memory.go

```diff
@@ -0,0 +1,54 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+var allocations = make(map[uintptr][]byte)
+
+func allocate(size uint32) (uint32, func()) {
+	fmt.Printf("allocating %v bytes", size)
+
+	return Write(make([]byte, size))
```

Comment:
I wonder if we can reuse the same slice, so we don't allocate a new one for each command.

Reply:
If we reuse the slice, and need to resize it (because a larger command arrived), then the resize operation will allocate a new underlying array (or no?). We can also pre-allocate a large enough slice, but I'm not sure how much would that help.

Reply:
If the slice has the maximum size we need (size of the biggest command), we should be good. If the slice is resliced to a smaller size, the backing array is kept the same and no allocation happens. Only if we try to expand the slice and the underlying array doesn't have more capacity, then a new array is allocated ([example](https://go.dev/play/p/DAhWksp0huE)).

Btw, how can we know the size of the command in advance? I imagine we would always have to prepare for the largest command we can receive. Also, can we make all commands a constant size?

Reply:
I'm thinking that commands could be as simple as an enum, a convention where `0=configure`, `1=open` etc.

Reply:
I imagine that after receiving the command we would then decide what to do first, for some commands it would be to fetch the parameters, for others it would just execute whatever was expected (e.g. `Teardown` requires no parameters). This would make it easier to allocate what's needed, since each command would decide what to allocate.

Did you have a different approach in mind?

Reply:
@lovromazgon  If we first ask for the command type and then for the parameters, then we still need to know the size of the serialized parameters. What we could do is to "combine" the command type and the serialized parameters' size into a single number (using bitwise operations).

If we want to allocate only the memory that is needed, then I believe we will always need two calls per command (the WASM module is driving the process, so it needs to ask for the needed memory and then the command). That's something that I wanted to avoid, which is why the code currently pre-allocates some memory.

If we want to do away with allocation, then I think the suggestion to simply prepare for the biggest command should work always. Responses should fit into those as well (the processed records).

Reply:
I don't think we can get around two round-trips, one to ask for the command/allocation size and the second one to populate the pointer, specifically because we can't know the size of the record. For all we know the record is multiple gigabytes (there's no limit currently 🙃), we don't want to pre-allocate that. And fetching a record is actually the hot path that we care about. We could optimize the other commands to a degree, but it's not really worth it, all other commands are going to be executed a single time.

So I'm still leaning towards fetching the command + allocation size in one call and then a second call to populate the parameters for that command (i.e. record).

Reply:
Definitely, pre-allocating won't work equally well for all the cases. What I think is that we're good for now, because:
* in practice, we only support individual records as input
* the individual records will, very often, be passed to a standalone connector, so they'll be limited to 4 MB anyway.

With that, IMHO, it makes sense to have the allocation as it is now and we can improve it later on (I created an issue for that a couple of weeks ago).

Reply:
I'm worried about the records that will fail if we pre-allocate too little, right now the record size limit is 1kB for a WASM processor, right?

How about this - I see that Conduit responds with the allocated size if it was successful, while if the size was not enough it returns a number representing an error. How about Conduit still responds with the size it would need to allocate the command, then the plugin can increase the allocation and ask for the command again? Then we can still get away with a single call and be sure we handle bigger records as well. In the future we can even be smart about the pre-allocation and adjust it based on the average size of the last N records/commands.

Reply:
Definitely! We can tweak the pre-allocated size to something more. The 1kB is from the very first commits, we can have it as maybe 4 MB?

As for the suggestion you described, that's basically #6 if I understood correctly?

Reply:
Lol yes, that's essentially what I meant, I could have saved a few brain cycles if only I read that issue 🙃

Reply:
I think 4MB is going to be more than the average record that we push around, I think 1kB might be a good starting point, as long as we retry if it's more than that. Also in a fast pipeline it will create huge pressure on the garbage collector unless we reuse the byte slice. We can merge this and I can work on #6 next week 👍

Reply:
> Lol yes, that's essentially what I meant, I could have saved a few brain cycles if only I read that issue 🙃

That's also my bad, I should have just pasted a link to the issue in the code!